code/modules/overmap/ftl_shunt/core.dm

```diff
@@ -0,0 +1,486 @@
+
+			uhoh = TRUE //let the fuckening begin!
+
+			for(var/obj/machinery/power/apc/A in SSmachines.machinery) //Effect One: shut down power across the ship.
+				if(!(A.z in ftl_computer.linked.map_z))
+					continue
+				A.energy_fail(rand(100,120))
+
+			for(var/mob/living/carbon/human/H in view(7)) //scary text if you're in view, because you're fucked now boy.
+				to_chat(H, SPAN_DANGER("The light around [src] warps before it emits a flash of incredibly bright, searing light!"))
+				H.flash_eyes(FLASH_PROTECTION_NONE)
+
+
+	ftl_announcement.Announce(announcetxt, "FTL Shunt Management System", new_sound = sound('sound/misc/ftlsiren.ogg'))
+
+	if(uhoh)
+		new /obj/singularity/(get_turf(src))
+
+//Returns status to ftl computer.
+/obj/machinery/ftl_shunt/core/proc/get_status()
+	if(stat & (BROKEN|NOPOWER))
+		return FTL_STATUS_OFFLINE
+	if(cooldown)
+		return FTL_STATUS_COOLDOWN
+	else
+		return FTL_STATUS_GOOD
+
+/obj/machinery/ftl_shunt/core/proc/get_fuel(var/list/input)
+	var/total_fuel_joules
+
+	for(var/obj/machinery/ftl_shunt/fuel_port/F in input)
+		total_fuel_joules += F.get_fuel_joules()
+
+	return total_fuel_joules
+
+/obj/machinery/ftl_shunt/core/proc/get_fuel_maximum(var/list/input)
+	var/max_fuel_joules = 0
+	for(var/obj/machinery/ftl_shunt/fuel_port/F in input)
+		max_fuel_joules += F.get_fuel_joules(TRUE)
+
+	return max_fuel_joules
+
+obj/machinery/ftl_shunt/core/proc/fuelpercentage()
+	if(!length(fuel_ports))
+		return 0
+	var/fuel_max = get_fuel_maximum(fuel_ports)
+	if(fuel_max == 0)
+		return 0
+	return round(100.0*get_fuel(fuel_ports)/fuel_max, 0.1)
+
+
+/obj/machinery/ftl_shunt/core/proc/use_fuel(var/joules_req)
+	var/joules_per_port
+	var/avail_fuel = get_fuel(fuel_ports)
+	var/list/fueled_ports = list()
+	var/ports_used
+
+	if(joules_req > avail_fuel) //Not enough fuel in the system.
+		return FALSE
+
+	for(var/obj/machinery/ftl_shunt/fuel_port/F in fuel_ports)
+		if(F.has_fuel())
+			fueled_ports += F
+
+	joules_per_port = (joules_req / length(fueled_ports))
+
+	for(var/obj/machinery/ftl_shunt/fuel_port/F in fueled_ports)
+		if(F.use_fuel_joules(joules_per_port))
+			ports_used++
+
+	if(ports_used == length(fueled_ports))
+		return TRUE
+
+/obj/machinery/ftl_shunt/core/proc/get_charge_time()
+	return (ftl_computer.linked.vessel_mass * CHARGE_TIME_PER_TON) * time_multiplier
+
+/obj/machinery/ftl_shunt/core/Process()
+	if((stat & (BROKEN|NOPOWER)) & charging)
+		cancel_shunt()
+
+	if(!length(fuel_ports))
+		find_ports()
+
+	if(charging)
+
+		if(sabotaged && !sabotaged_warn)
+			for(var/mob/living/carbon/human/H in world) //Give engineers a hint that something might be very, very wrong.
+				if(!(H.z in ftl_computer.linked.map_z))
+					continue
+				if(H.skill_check(SKILL_ENGINES, SKILL_EXPERT))
+					to_chat(H, SPAN_DANGER("The deck vibrates with a harmonic that sets your teeth on edge and fills you with dread."))
+			sabotaged_warn = TRUE
+
+		if(use_power != POWER_USE_ACTIVE)
+			update_use_power(POWER_USE_ACTIVE)
```

Comment:
Why do this every tick? This entire block feels like it should be done on charge start, not in Process.